clients/esp32/client.ino

```diff
@@ -0,0 +1,272 @@
+        Serial.println("Connected to Wi-Fi");
+    } else {
+        Serial.println("Failed to connect to Wi-Fi. Check credentials.");
+    }
+}
+
+void setup() {
+  // put your setup code here, to run once:
+  Serial.begin(115200);
+  Serial.println("setting up wifi");
+  WiFi.mode(WIFI_STA);
+  connectToWifi(ssid, password);
+  Serial.println("connected to wifi");
+  Serial.println("let's try playing some audio");
+
+  i2s_config_t i2s_config = {
+    .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_RX | I2S_MODE_TX),
+    .sample_rate = 24000, // Adjust this according to your audio file's sample rate
+    .bits_per_sample = I2S_BITS_PER_SAMPLE_32BIT, // Adjust this too
+    .channel_format = I2S_CHANNEL_FMT_ONLY_LEFT, // Mono
+    .communication_format = I2S_COMM_FORMAT_I2S,
+    .intr_alloc_flags = 0, // Default interrupt priority
+    .dma_buf_count = 4,
+    .dma_buf_len = 1024,
+    .use_apll = false,
+    .tx_desc_auto_clear = true, // Auto clear tx descriptor on underflow
+    .fixed_mclk = 0
+  };
+
+  i2s_pin_config_t pin_config = {
+    .bck_io_num = 16, // BCLK
+    .ws_io_num = 9, // LRCK
+    .data_out_num = 15, // Speaker output
+    .data_in_num = 17 // Microphone input
+  };
+
+  // Install and start I2S driver with your new pin configuration
+  i2s_driver_install(i2s_num, &i2s_config, 0, NULL);
+  i2s_set_pin(i2s_num, &pin_config);  
+
+  i2s_set_sample_rates(I2S_NUM_0, 22050); // Set the sample rate for your audio file.
+  const char* audioURL = "http://quack.uberduck.ai/aec-cartoon.wav";
+  // playAudioFromURL(audioURL);
+
+  webSocket.setCACert(rootCACertificate);
+
+
+  i2s_set_sample_rates(I2S_NUM_0, 16000);
+  Serial.println("switched to 16khz sample rate");
+  
+
+  HTTPClient http;
+  http.begin("https://9df3b80e5aae.ngrok.app/status");
+  int httpCode = http.GET();
+  if (httpCode == HTTP_CODE_OK) {
+    auto stream = http.getStream();
+    uint8_t buffer[1024];
+    size_t bytesRead = 0;
+
+    // Loop to read from the stream
+    while ((bytesRead = stream.readBytes(buffer, sizeof(buffer))) > 0) {
+      // Print each chunk of data as it's read
+      Serial.write(buffer, bytesRead);
+    }
+    
+  } else {
+    Serial.printf("HTTP request failed: %s\n", http.errorToString(httpCode).c_str());
+  }
+  http.end();
+  Serial.println("connecting to websocket.");
+  webSocket.onMessage(onMessageCallback);
+  webSocket.onEvent(onEventsCallback);
+  String session_id = generateRandomUniqueString();
+  webSocket.connect(
+    "wss://9df3b80e5aae.ngrok.app:443/audio/response?session_id=" + 
+    session_id + 
+    "&record=true" + 
+    "&input_audio_format=int16&output_sample_rate=16000"
+  );
+  webSocket.ping();
+}
+
+void loop() {
+  webSocket.poll();
+
+  unsigned long currentMillis = millis();
+  if (currentMillis - previousPing > pingInterval) {
+    webSocket.ping();
+    previousPing = currentMillis;
+  }
+
+  esp_err_t err = i2s_read(
+      I2S_NUM_0,
+      (void *) i2s_read_buff_int32,
```

Comment:
can we make this i2s_read_buff_int16() and then not need to convert it on line 248?

Reply:
no - the i2s device on my hardware required int32 audio.